python/sqlflow_submitter/pai/submitter.py

```diff
@@ -0,0 +1,301 @@
+        submit_tables = "%s,%s" % (submit_tables, val_table)
+    output_tables = ""
+    if res_table != "":
+        table = max_compute_table_url(res_table)
+        output_tables = "-Doutputs=%s" % table
+
+    # NOTE(typhoonzero): use - DhyperParameters to define flags passing OSS credentials.
+    # TODO(typhoonzero): need to find a more secure way to pass credentials.
+    cmd = ("pai -name tensorflow1150 -project algo_public_dev "
+           "-DmaxHungTimeBeforeGCInSeconds=0 -DjobName=%s -Dtags=dnn "
+           "-Dscript=%s -DentryFile=%s -Dtables=%s %s -DhyperParameters=\"%s\"") % (
+        job_name, tarball, entry_file, submit_tables, output_tables, params_file)
+
+    # format the oss checkpoint path with ARN authorization.
+    oss_checkpoint_configs = os.getenv("SQLFLOW_OSS_CHECKPOINT_DIR")
+    if oss_checkpoint_configs == "":
+        raise "need to configure SQLFLOW_OSS_CHECKPOINT_DIR when submitting to PAI"
+    oss_json_configs = json.loads(oss_checkpoint_configs)
+    curr_project_oss = oss_json_configs[project]
+    if project not in curr_project_oss:
+        raise "project %s not configured in SQLFLOW_OSS_CHECKPOINT_DIR" % project
+
+    arn_splited = curr_project_oss.split("?")
+    if len(arn_splited) != 2:
+        raise "need to configure SQLFLOW_OSS_CHECKPOINT_DIR when submitting to PAI"
+    arn = arn_splited[1]
+    ossURI = get_oss_model_url(oss_model_path)
+    oss_checkpoint_path = "%s/?%s" % (ossURI, arn)
+    cmd = "%s -DcheckpointDir='%s'" % (cmd, oss_checkpoint_path)
+
+    if cluster_config["worker"]["count"] > 1:
+        cmd = "%s -Dcluster=\"%s\"" % (cmd, cf_quote)
+    else:
+        cmd = "%s -DgpuRequired='%d'" % (cmd, cluster_config["worker"]["gpu"])
+    return cmd
+
+
+def prepare_archive(cwd, conf, project, estimator, model_name,
+                    train_tbl, val_tbl, model_save_path, train_params):
+    """package needed resource into a tarball"""
+    create_pai_hyper_param_file(cwd, PARAMS_FILE, model_save_path)
+
+    with open(path.join(cwd, TRAIN_PARAMS_FILE), "w") as param_file:
+        pickle.dump(train_params, param_file)
+
+    with open(path.join(cwd, "requirements.txt"), "w") as require:
+        require.write(TF_REQUIREMENT)
+    copy_python_package("sqlflow_submitter", cwd)
+    copy_python_package("sqlflow_models", cwd)
+    copy_custom_package(estimator, cwd)
+
+    os.execl("tar", "czf", JOB_ARCHIVE_FILE, "./sqlflow_submitter",
+             "./sqlflow_models", "requirements.txt", TRAIN_PARAMS_FILE)
+
+
+def save_model_to_sqlfs(datasource, model_oss_path, model_name):
+    # (TODO: save model to sqlfs)
+    pass
+
+
+# (TODO: lhw) adapt this interface after we do feature derivation in Python
+def submit_pytf_train(datasource, estimator, select, validation_select,
+                      model_params, model_name, pre_trained_model, **train_params):
+    """
+    This function submit PY-TF train task to PAI platform
+
+    Parameters
+    ----------
+    datasource: string
```

Comment:
Use this style to format docstrings? https://github.com/google/tangent/blob/master/tangent/ast.py
This is also what TensorFlow's docstring style.

Reply:
done